File: rust/lib/src/execution/test_suite_executor.rs
use std::{borrow::Borrow, collections::{HashMap}, thread::sleep, time::Duration};
use anyhow::{Context, Result, anyhow};
use log::debug;
use tokio::runtime::Runtime;
use tonic::{Request, transport::{Channel}};

use crate::{core_api_bindings::api_container_api::{SuiteAction, TestMetadata, TestSuiteMetadata, api_container_service_client::ApiContainerServiceClient, suite_metadata_serialization_service_client::SuiteMetadataSerializationServiceClient, suite_registration_service_client::SuiteRegistrationServiceClient, test_execution_service_client::TestExecutionServiceClient}, testsuite::testsuite::TestSuite};

use super::test_suite_configurator::TestSuiteConfigurator;

const GRPC_SERVER_STOP_GRACE_PERIOD: Duration = Duration::from_secs(5);
const MAX_CONNECTION_ATTEMPTS: u32 = 20;
const TIME_BETWEEN_CONNECTION_RETRIES: Duration = Duration::from_millis(500);

pub struct TestSuiteExecutor {
    kurtosis_api_socket: Option<String>, // Can be empty if testsuite is in metadata-providing mode
    log_level: String,
    params_json: String,
	configurator: Box<dyn TestSuiteConfigurator>,
}

impl TestSuiteExecutor {
    pub fn new(kurtosis_api_socket: Option<String>, log_level: &str, params_json: &str, configurator: Box<dyn TestSuiteConfigurator>) -> TestSuiteExecutor {
        return TestSuiteExecutor{
            kurtosis_api_socket,
            log_level: log_level.to_owned(),
            params_json: params_json.to_owned(),
			configurator,
        };
    }

	pub fn run(&self) -> Result<()> {
		self.configurator.set_log_level(&self.log_level)
			.context("An error occurred setting the loglevel before running the testsuite")?;

		let suite = self.configurator.parse_params_and_create_suite(&self.params_json)
			.context("An error occurred parsing the suite params JSON and creating the testsuite")?;

		let mut api_container_service: Option<ApiContainerServiceClient<Channel>> = None;
		match self.kurtosis_api_socket {
			Some(kurtosis_api_socket_str) => {
				let url = format!("http://{}", kurtosis_api_socket_str);
				// TODO SECURITY: Use HTTPS to ensure we're connecting to the real Kurtosis API servers
				let endpoint = Channel::from_shared(url)
					.context(format!("An error occurred creating the endpoint to Kurtosis API socket '{}'", kurtosis_api_socket_str))?;

				let async_runtime = Runtime::new()
					.context("An error occurred creating the Tokio async runtime")?;

				// Sometimes the API container is still in the process of starting, so we retry the channel connection a few times
				let channel: Channel;
				let mut connection_attempts: u32 = 0;
				loop {
					if connection_attempts >= MAX_CONNECTION_ATTEMPTS {
						return Err(
							anyhow!(
								"Failed to connect to API container, even after {} retries spaced {}ms apart",
								MAX_CONNECTION_ATTEMPTS,
								TIME_BETWEEN_CONNECTION_RETRIES.as_millis(),
							)
						);
					}
					let channel_or_err = async_runtime.block_on(endpoint.connect());
					match channel_or_err {
						Ok(returned_channel) => {
							channel = returned_channel;
							break;
						}
						Err(err) => {
							debug!(
								"The following error occurred connecting to the API container; retrying in {}ms: {}", 
								TIME_BETWEEN_CONNECTION_RETRIES.as_millis(),
								err.to_string()
							);
						}
					}
					sleep(TIME_BETWEEN_CONNECTION_RETRIES);
					connection_attempts += 1;
				}

				let api_container_client_channel = channel.clone(); // This *seems* weird to clone a channel, but this is apparently how Tonic wants it
				api_container_service = Some(ApiContainerServiceClient::new(api_container_client_channel));
			},
			None => {},
		}

		// TODO

		return Ok(());
	}
}

File: golang/lib/execution/test_suite_executor.go
package execution

import (
	"github.com/kurtosis-tech/kurtosis-client/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-libs/golang/lib/rpc_api/bindings"
	"github.com/kurtosis-tech/kurtosis-libs/golang/lib/rpc_api/rpc_api_consts"
	"github.com/kurtosis-tech/minimal-grpc-server/server"
	"github.com/palantir/stacktrace"
	"google.golang.org/grpc"
	"time"
)

const (
	grpcServerStopGracePeriod = 5 * time.Second
)

type TestSuiteExecutor struct {
	kurtosisApiSocket string  // Can be empty if the testsuite is in metadata-providing mode
	logLevelStr string
	paramsJsonStr string
	configurator TestSuiteConfigurator
}

func NewTestSuiteExecutor(kurtosisApiSocket string, logLevelStr string, paramsJsonStr string, configurator TestSuiteConfigurator) *TestSuiteExecutor {
	return &TestSuiteExecutor{kurtosisApiSocket: kurtosisApiSocket, logLevelStr: logLevelStr, paramsJsonStr: paramsJsonStr, configurator: configurator}
}

func (executor TestSuiteExecutor) Run() error {
	if err := executor.configurator.SetLogLevel(executor.logLevelStr); err != nil {
		return stacktrace.Propagate(err, "An error occurred setting the loglevel before running the testsuite")
	}

	suite, err := executor.configurator.ParseParamsAndCreateSuite(executor.paramsJsonStr)
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred parsing the suite params JSON and creating the testsuite")
	}

	var apiContainerService kurtosis_core_rpc_api_bindings.ApiContainerServiceClient = nil
	if executor.kurtosisApiSocket != "" {
		// TODO SECURITY: Use HTTPS to ensure we're connecting to the real Kurtosis API servers
		conn, err := grpc.Dial(executor.kurtosisApiSocket, grpc.WithInsecure())
		if err != nil {
			return stacktrace.Propagate(
				err,
				"An error occurred creating a connection to the Kurtosis API server at '%v'",
				executor.kurtosisApiSocket,
			)
		}
		defer conn.Close()

		apiContainerService = kurtosis_core_rpc_api_bindings.NewApiContainerServiceClient(conn)
	}

	testsuiteService := NewTestSuiteService(suite, apiContainerService)
	testsuiteServiceRegistrationFunc := func(grpcServer *grpc.Server) {
		bindings.RegisterTestSuiteServiceServer(grpcServer, testsuiteService)
	}

	testsuiteServer := server.NewMinimalGRPCServer(
		rpc_api_consts.ListenPort,
		rpc_api_consts.ListenProtocol,
		grpcServerStopGracePeriod,
		[]func(desc *grpc.Server) {
			testsuiteServiceRegistrationFunc,
		},
	)
	if err := testsuiteServer.Run(); err != nil {
		return stacktrace.Propagate(err, "An error occurred running the testsuite server")
	}

	return nil
}
